core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala

```diff
@@ -367,21 +368,55 @@ object Protocol {
       !metadata.configuration.contains(DeltaConfigs.CREATE_TABLE_IGNORE_PROTOCOL_DEFAULTS.key),
       "Should not have the table property " +
         s"${DeltaConfigs.CREATE_TABLE_IGNORE_PROTOCOL_DEFAULTS.key} stored in table metadata")
-    val (readerVersion, writerVersion, enabledFeatures) =
+    val (readerVersion, writerVersion, minRequiredFeatures) =
       minProtocolComponentsFromAutomaticallyEnabledFeatures(spark, metadata)
 
     // Increment the reader and writer version to accurately add enabled legacy table features
     // either to the implicitly enabled table features or the table feature lists
     val required = Protocol(
         readerVersion.max(current.minReaderVersion), writerVersion.max(current.minWriterVersion))
-        .withFeatures(enabledFeatures)
+        .withFeatures(minRequiredFeatures)
     if (!required.canUpgradeTo(current)) {
+      // When the current protocol does not satisfy metadata requirement, some additional features
+      // must be supported by the protocol. We assert those features can actually perform the
+      // auto-update.
+      assertMetadataTableFeaturesAutomaticallySupported(
```

Comment:
The current function now changes from just calculating a new protocol (before) to calculating a new protocol _only if it can be automatically upgraded_. The name `checkProtocolRequirements` is pretty terrible for both, and the function documentation also says nothing about checking for exceptions. I would propose renaming the current function to e.g. `upgradeProtocolFromMetadataForExistingTable` and updating the function documentation to say what it actually does (apply automatic protocol upgrades and throw an error if a non-automatic protocol upgrade is required).

Reply:
Done! I also moved those `assert`s into a separate method.